### PR TITLE
VSCode Guide Developer Documentation

### PR DESCRIPTION
**Pull request by Pasarus**
**Description of work.**
A Developer guide on how best to work with VSCode aimed at new started mostly centred around operation on systems not operating on Windows.

**To test:**
- Build the dev-docs-html target
- Open "Build/Directory/dev-docs/html/index.html" in a browser
- Navigate down to Tools
- Click on VSCode
- Check that the page looks fine and the guide makes sense

*There is no associated issue.*

*This does not require release notes* as it is a developer doc

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: dev-docs/source/VSCode.rst [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+                "ignoreFailures": true
+            }
+        ]
+    }
+
+Keybindings
+-----------
+
+To get a list of all of possible keybindings the open your command line
+Ctrl+Shift+P and search for "Help: Keyboard Shortcuts Reference" and hit Enter.
```
Comment (TTitcombe) [inline dev-docs/source/VSCode.rst]:
```suggestion
and search for "Help: Keyboard Shortcuts Reference" and hit Enter.
```

**Review by TTitcombe — CHANGES_REQUESTED** [review]
Clear and helpful. Just a couple of minor comments

File: dev-docs/source/VSCode.rst [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+- Search for "Auto Save"
+- Select the auto save type that is wanted (recommend `onFocusChange`)
+
+Building
+--------
+This is where VSCode comes into it's own as a versatile environment for programming
+in any language, because of this it has a Tasks list. These tasks are saved in
+tasks.json inside the .vscode folder now present in any opened folder/workspace.
+
+- To start creating a task open the command line = Ctrl+Shift+P
```
[the remark this answers is not in the document]
Comment (Pasarus) [reply dev-docs/source/VSCode.rst]: I think it is important to have this keybinding at this point as the dev docs haven't discussed it yet. However, I will say that it works on Linux/Windows but not MacOS.

File: dev-docs/source/VSCode.rst [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+
+Recommended
+-----------
+- `Clang-Format <https://marketplace.visualstudio.com/items?itemName=xaver.clang-format>`_
+    - Allows auto formatting using the clang format version installed on your machine
+- `CMake <https://marketplace.visualstudio.com/items?itemName=twxs.cmake>`_
+    - Allows language highlighting and auto fill for CMake
+- `Doxygen Document Generator <https://marketplace.visualstudio.com/items?itemName=cschlosser.doxdocgen>`_
+    - Once a function is written with this extension it can generate parts of the doxygen comments itself
+- `GitLens <https://marketplace.visualstudio.com/items?itemName=eamodio.gitlens>`_
```
[the remark this answers is not in the document]
Comment (Pasarus) [reply dev-docs/source/VSCode.rst]: I don't believe the extension to be that valuable as all good git applications and command line can perform these tasks and will definitely also be in use.

**Review by TTitcombe — DISMISSED** [review]

File: dev-docs/source/VSCode.rst [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+==================
+- Run VSCode
+- Click File->Open Folder
+- Navigate to your Mantid source directory and select it
+- Install the required extensions (see below)
+- For code editing you are good to go!
+
+Extensions
+==========
+All Extensions have been tested working on Ubuntu 18.04 LTS and Ubuntu 19.04,
```
Comment (DanNixon) [inline dev-docs/source/VSCode.rst]:
```suggestion
All Extensions have been tested working on Ubuntu 18.04 and Ubuntu 19.04,
```

File: dev-docs/source/VSCode.rst [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+    - Very powerful source control assistant
+- `Visual Studio Intellicode <https://marketplace.visualstudio.com/items?itemName=VisualStudioExptTeam.vscodeintellicode>`_
+    - This improves your code suggestions in certain languages
+
+Features
+========
+
+Auto Formatting
+---------------
+With a formatter installed as an extension it is recommended to use Clang-Format. It
```
Comment (DanNixon) [inline dev-docs/source/VSCode.rst]: This sentence is a bit odd. Can it be rephrased or removed?

File: dev-docs/source/VSCode.rst [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+- Now to build with this task open the command line again
+- Type "Task"
+- Select "Tasks: Run Build Task"
+
+
+Debugging and Launching
+-----------------------
+Debugging is similar to Building in the sense that you complete a task that has been
+defined. For Debugging and Launching all of these 'Tasks' are stored in the
+launch.json alongside the tasks.json in the .vscode folder.
```
Comment (DanNixon) [inline dev-docs/source/VSCode.rst]:
```suggestion
``launch.json`` alongside the ``tasks.json`` in the ``.vscode`` folder.
```

(I think double backtick is the correct syntax for raw text)

File: dev-docs/source/VSCode.rst [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+
+.. code-block:: javascript
+
+    {
+        "version": "2.0.0",
+        "tasks": [
+            {
+                "label": "Build Mantid",
+                "type": "shell",
+                "command": "ninja",
```
Comment (DanNixon) [inline dev-docs/source/VSCode.rst]:
```suggestion
                "command": "cmake --build .",
```

This makes the Ninja assumption above irrelevant.

**Review by DanNixon — CHANGES_REQUESTED** [review]
A few style points.

There are also a lot of cases where random words mid sentence seem to be capitalized for no reason and filenames not in raw/unformatted/code format.

File: dev-docs/source/VSCode.rst (hunk 7 above)
Comment (Pasarus) [reply dev-docs/source/VSCode.rst]: I don't agree with this change, as we should be using ninja as our build system on MacOS or Linux.

Comment (DanNixon) [reply dev-docs/source/VSCode.rst]: The above call will use Ninja if that's the generator you specified. `cmake --build .` will always work so should be preferred here.

File: dev-docs/source/VSCode.rst [hunk 8]
```diff
@@ -107,8 +107,9 @@ switched out with the command and various args for whatever you have generated w
             {
                 "label": "Build Mantid",
                 "type": "shell",
-                "command": "ninja",
+                "command": "cmake",
                 "args": [
+                    "--build",
```
Comment (DanNixon) [inline dev-docs/source/VSCode.rst]: You do not need the `-C` here. The proper syntax is `cmake --build [path to build]`.

**Review by DanNixon — CHANGES_REQUESTED** [review]

**Comment by Pasarus** [comment]
@DanNixon Is this ready to be merged yet?

File: dev-docs/source/VSCode.rst [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+- `Visual Studio Intellicode <https://marketplace.visualstudio.com/items?itemName=VisualStudioExptTeam.vscodeintellicode>`_
+    - This improves your code suggestions in certain languages
+
+Features
+========
+
+Auto Formatting
+---------------
+With a formatter installed as an extension it is possible to format any open file in VSCode.
+It is recommended to use `Clang-Format <https://marketplace.visualstudio.com/items?itemName=xaver.clang-format>`_.
```
Comment (DanNixon) [inline dev-docs/source/VSCode.rst]: We require `clang-format`.

**Review by DanNixon — CHANGES_REQUESTED** [review]

File: dev-docs/source/VSCode.rst (hunk 9 above)
Comment (Pasarus) [reply dev-docs/source/VSCode.rst]: I understand that, however turning on auto-formatting in VSCode is optional, there are other ways to format it, I also was just recommending one of the Clang-format extensions.

Comment (DanNixon) [reply dev-docs/source/VSCode.rst]: That should be more obvious, right now it reads "formatting is a thing, you should probably do it with clang-format".

Comment (Pasarus) [reply dev-docs/source/VSCode.rst]: I have made a change to how the test reads.

**Review by DanNixon — APPROVED** [review]

**Pull request merged**